design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,474 @@
+metadata:
+  name: db-secret
+  labels:
+    configurable-key: env-prod # label name can be configurable
+  type: Opaque
+stringData:
+  conjur-map: |-
+    username: secrets-accessors/my_username
+```
+
+These labels do not have to be Conjur-specific and the customer would decide which label to attach to the K8s secrets. They would define which secrets to iterate over by `K8S_SECRET_LABEL` env variable in their Secrets Provider manifest. The Secrets Provider will perform a search for secrets with that label key-value pair against the [K8s](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#secret-v1-core)/[OC](https://docs.openshift.com/container-platform/4.4/rest_api/index.html#secret-v1-core) API and search for the "conjur_map" on the results.
+
+Note that just because the customer added a label for us to filter over doesn't mean that all secret entries in the K8s Secret are stored in Conjur but it gives us a way to perform a more focused search for "conjur-map" in the K8s Secrets.
+
+Secrets Provider Manifest with the `K8S_SECRET_LABEL` environment variable:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: secrets-provider-deployment
+spec:
+  replicas: 1
+  template:
+    spec:
+      serviceAccountName: sa-1
+      containers:
+      - image: secrets-provider:latest
+        name: cyberark-secrets-provider
+   ...
+      env:
+        - name: K8S_SECRET_LABELS
+          value: 'configurable-key:env-prod,environment:prod'
+```
+
+These labels will have an OR inclusive relationship, not an AND relationship. A customer can also write '*' to imply fetching all K8s secrets in the namespace. If K8S_SECRET_LABELS is not supplied, by default, the Secrets Provider will fetch all K8s Secrets in the namespace.
+
+To read the labels set on K8s Secrets, the Service Account used for the Secrets Provider will need "list" privileges on K8s Secrets resources.
+
+The Service Account / Role / RoleBinding will be defined as the following:
+
+```yaml
+# Service Account definition 
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: conjur-sa-1
+
+# Role definition
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  namespace: app-namespace
+  name: secrets-access
+rules:
+- apiGroups: [""]
+  resources: ["secrets"]
+  verbs: ["get", "update", "list"] # new privilege here
+
+# RoleBinding definition to associate the SA with the Role
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: secrets-access-binding
+  namespace: app-namespace
+subjects:
+  - kind: ServiceAccount
+    name: conjur-sa-1
+    apiGroup: rbac.authorization.k8s.io
+roleRef:
+  kind: Role
+  name: secrets-access
+  apiGroup: rbac.authorization.k8s.io
+```
+
+#### Code changes
+
+Although we will now offer "K8S_SECRET_LABELS" for easy filtering on K8s Secrets, we will still support the "K8S_SECRET" environment variable if the customer prefers to list the K8s Secrets they need values for from Conjur. 
```

Comment:
what are the code changes here? please define the places in the code that will change during this effort.

Reply:
The code changes will take place in `config.go` file

We would remove `K8S_SECRETS` from the required env and add a condition that either `K8S_SECRETS` or `K8S_SECRETS_LABEL` needs to be present
```
for _, envvar := range []string{
    "MY_POD_NAMESPACE",
    "K8S_SECRETS",
    "SECRETS_DESTINATION",
} {
    if os.Getenv(envvar) == "" {
        return nil, log.RecordedError(messages.CSPFK004E, envvar)
    }
}
```
If `K8S_SECRETS_LABEL`, then there is added work that needs to be done of making an extra call to K8s API to get the secrets with the correct label. 

This will determine the value for `requiredK8sSecrets` in `Config` object
```
return &Config{
    PodNamespace:       podNamespace,
    RequiredK8sSecrets: requiredK8sSecrets,
    StoreType:          storeType,
}, nil
```

Do you suggest actually changing the code and adding it to doc?

Reply:
not changing the code, just specifying the places that the code will change so we can grasp it. It seems that the places that will change in this effort is those you specify above and also other places (batch retrieval for example). This section is to list all those places that will change.

If you think this section has no value please challenge it.

Reply:
added!